src/main/scala/org/broadinstitute/hail/driver/Join.scala

```diff
@@ -0,0 +1,73 @@
+package org.broadinstitute.hail.driver
+
+import org.broadinstitute.hail.utils._
+import org.broadinstitute.hail.variant._
+import org.kohsuke.args4j.{Option => Args4jOption}
+
+object Join extends Command {
+
+  class Options extends BaseOptions {
+    @Args4jOption(required = true, name = "-r", aliases = Array("--right"), usage = ".vds file to join on the right")
+    var right: String = _
+  }
+
+  def newOptions = new Options
+
+  def name = "join"
+
+  def description = "Join datasets, inner join on variants, concatenate samples, variant and global annotations from left"
+
+  def supportsMultiallelic = true
+
+  def requiresVDS = true
+
+  def run(state: State, options: Options): State = {
+    val left = state.vds
+    val right = VariantSampleMatrix.read(state.sqlContext, options.right)
+
+    if (left.wasSplit != right.wasSplit) {
+      fatal(
+        s"""cannot join split and unsplit datasets
+            |  left was split: ${ left.wasSplit }
+            |  light was split: ${ right.wasSplit }""".stripMargin)
+    } else if (left.saSignature != right.saSignature) {
+      fatal(
+        s"""cannot join datasets with different sample schemata
```

Comment:
I'm inclined to make the join as simple as possible.  You can do that with an `annotatesamples` easily enough.